modules/apps/fragment/fragment-impl/src/main/java/com/liferay/fragment/internal/processor/PortletRegistryImpl.java

```diff
@@ -189,50 +192,82 @@ public void writePortletPaths(
-		String portletName = MapUtil.getString(
-			properties, "javax.portlet.name");
-
-		_portletNames.put(alias, portletName);
-
-		Bundle bundle = FrameworkUtil.getBundle(jxPortlet.getClass());
-
-		BundleContext bundleContext = bundle.getBundleContext();
-
-		_serviceRegistrations.put(
-			alias,
-			bundleContext.registerService(
-				PortletAliasRegistration.class,
-				new PortletAliasRegistration() {
-				},
-				HashMapDictionaryBuilder.<String, Object>put(
-					"com.liferay.fragment.entry.processor.portlet.alias", alias
-				).build()));
-	}
+	@Activate
+	protected void activate(BundleContext bundleContext)
+		throws InvalidSyntaxException {
+
+		String filterString = StringBundler.concat(
+			"(&(com.liferay.fragment.entry.processor.portlet.alias=*)",
+			"(objectClass=", javax.portlet.Portlet.class.getName(), "))");
+
+		_serviceTracker = new ServiceTracker<>(
+			bundleContext, bundleContext.createFilter(filterString),
+			new ServiceTrackerCustomizer
+				<javax.portlet.Portlet, javax.portlet.Portlet>() {
+
+				@Override
+				public javax.portlet.Portlet addingService(
+					ServiceReference<javax.portlet.Portlet> serviceReference) {
+
+					String alias = GetterUtil.getString(
+						serviceReference.getProperty(_keyAlias));
+
+					_portletNames.put(
+						alias,
+						GetterUtil.getString(
+							serviceReference.getProperty(_keyPortletName)));
+
+					_serviceRegistrations.put(
+						alias,
+						bundleContext.registerService(
+							PortletAliasRegistration.class,
+							new PortletAliasRegistration() {
+							},
+							MapUtil.singletonDictionary(_keyAlias, alias)));
+
+					return bundleContext.getService(serviceReference);
+				}
 
-	protected void unsetPortlet(
-		javax.portlet.Portlet jxPortlet, Map<String, Object> properties) {
+				@Override
+				public void modifiedService(
+					ServiceReference<javax.portlet.Portlet> serviceReference,
+					javax.portlet.Portlet jxPortlet) {
+				}
+
+				@Override
+				public void removedService(
+					ServiceReference<javax.portlet.Portlet> serviceReference,
+					javax.portlet.Portlet jxPortlet) {
+
+					String alias = GetterUtil.getString(
+						serviceReference.getProperty(_keyAlias));
 
-		String alias = MapUtil.getString(
-			properties, "com.liferay.fragment.entry.processor.portlet.alias");
-		String portletName = MapUtil.getString(
-			properties, "javax.portlet.name");
+					_portletNames.remove(
+						alias,
+						GetterUtil.getString(
+							serviceReference.getProperty(_keyPortletName)));
 
-		_portletNames.remove(alias, portletName);
+					bundleContext.ungetService(serviceReference);
 
-		ServiceRegistration<PortletAliasRegistration> serviceRegistration =
-			_serviceRegistrations.remove(alias);
+					ServiceRegistration<PortletAliasRegistration>
+						serviceRegistration = _serviceRegistrations.remove(
+							alias);
+
+					serviceRegistration.unregister();
+				}
+
+				private final String _keyAlias =
```

Comment:
@W2jx If you want to use these as constants to avoid writing the long string again and again, please make them `private static final` and put them in the outer class.